File: Parsing/StringInput.cs
using System;

namespace KontrolSystem.Parsing;

/// <summary>
///     Entire input for parsing is available as string.
/// </summary>
public readonly ref struct StringInput(ReadOnlySpan<char> source, string sourceName, int position = 0, int line = 1, int column = 1) {
    private readonly ReadOnlySpan<char> source = source;
    
    public char Current => source[0];

    public int Available => source.Length;

    public Position Position => new(sourceName, position, line, column);

    public int FindNext(Predicate<char> predicate) {
        for (var p = 0; p < source.Length; p++)
            if (predicate(source[p]))
                return p;

        return -1;
    }

    public bool Match(ReadOnlySpan<char> tag) => source.StartsWith(tag);

    public string Take(int count) {
        if (count == 0) return "";
        if (count > source.Length) throw new InvalidOperationException("Advance beyond eof");

        return source.Slice(0, count).ToString();
    }

    public StringInput Advance(int count) {
        if (count == 0) return this;
        if (count > source.Length) throw new InvalidOperationException("Advance beyond eof");

        var nextLine = line;
        var nextColumn = column;

        for (var p = 0; p < count; p++) {
            nextColumn++;
            if (source[p] == '\n') {
                nextLine++;
                nextColumn = 1;
            }
        }

        return new StringInput(source.Slice(count), sourceName, position + count, nextLine, nextColumn);
    }

    public override string ToString() {
        return source.ToString();
    }
}

File: Tools/vscode/to2-syntax/server/src/parser/string-input.ts
import { Input } from ".";

export class StringInput implements Input {
    private source: string;
    private position: number;

    constructor(source: string, position: number = 0) {
        this.source = source;
        this.position = position;
    }

    current(): string {
        return this.source[this.position];
    }

    available(): number {
        return this.source.length - this.position;
    }
    
    take(count: number): string {
        if (count === 0) return "";

        return this.source.substring(this.position, this.position + count);
    }

    findNext(predicate: (ch: string) => boolean): number {
        for(let p = this.position; p < this.source.length; p++) {
            if(predicate(this.source[p])) return p - this.position;
        }
        return -1;
    }
    
    advance(count: number): Input {
        return new StringInput(this.source, this.position + count);
    }
}